km/km_filesys.c

```diff
@@ -1800,25 +1800,37 @@ uint64_t km_fs_sendrecvmsg(km_vcpu_t* vcpu, int scall, int sockfd, struct msghdr
    }
    if (scall == SYS_sendmsg) {
       // translate sent file descriptors if any
-      for (struct cmsghdr* cmsg = CMSG_FIRSTHDR(msg); cmsg != NULL; cmsg = CMSG_NXTHDR(msg, cmsg)) {
-         if (cmsg->cmsg_type == SCM_RIGHTS) {
-            int guest_fd = *(int*)CMSG_DATA(cmsg);
-            int host_fd = km_fs_g2h_fd(guest_fd, NULL);
-            *(int*)CMSG_DATA(cmsg) = host_fd;
-            km_infox(KM_TRACE_FILESYS, "send guest fd %d as host %d", guest_fd, host_fd);
+      if (msg->msg_control != NULL) {
+         for (struct cmsghdr* cmsg = CMSG_FIRSTHDR(msg); cmsg != NULL; cmsg = CMSG_NXTHDR(msg, cmsg)) {
+            if (cmsg->cmsg_level == SOL_SOCKET && cmsg->cmsg_type == SCM_RIGHTS) {
+               int fdcount = (cmsg->cmsg_len - sizeof(struct cmsghdr)) / sizeof(int);
+	       km_infox(KM_TRACE_FILESYS, "sockfd %d: processing SCM_RIGHTS send chunk containing %d fd's", sockfd, fdcount);
+               int* guest_fd = (int*)CMSG_DATA(cmsg);
+	       for (int i = 0; i < fdcount; i++) {
+                  int host_fd = km_fs_g2h_fd(guest_fd[i], NULL);
+                  km_infox(KM_TRACE_FILESYS, "fd[%d] send guest fd %d as host fd %d", i, guest_fd[i], host_fd);
+                  guest_fd[i] = host_fd;
```

Comment:
This is altering the payloads memory but the fd isn't changing unless valgrind is being used.

We should use our own buffer for the altered control message.

Reply:
I don't follow the valgrind reference here. Files are always mapped one to one. Memory is mapped differently with valgrind

Reply:
You are right, I'm not thinking clearly.